File: lib/rollbar/dumps/builder.rb
require 'rollbar/dumps/gdb'

module Rollbar
  module Dumps
    class Builder
      attr_reader :proxy

      def initialize(proxy)
        @proxy = proxy
      end

      def to_h
        initial_response = proxy.start
        frames = build_frames
        klass, message = extract_klass_and_message(initial_response, frames)

        {
          'access_token' => Configuration.access_token,
          'data' => {
            'timestamp' => Time.now.to_i,
            'environment' => 'development',
            'level' => 'error',
            'language' => 'c',
            'server' => {
              'host' => Socket.gethostname
            },
            'notifier' => {
              'name' => 'rollbar-dumps',
              'version' => Rollbar::Dumps::VERSION
            },
            'body' => {
              'trace' => {
                'frames' => frames,
                'exception' => {
                  'class' => klass,
                  'message' => message
                }
              }
            }
          }
        }
      end

      private

      def extract_klass_and_message(response, frames)
        error_match = response.to_s.match(/Program terminated with signal (.+), (.*)\./)

        klass = error_match[1]
        message = "#{error_match[2]} (#{frames[-1][:filename]})"

        [klass, message]
      end

      def build_frames
        depth = proxy.stack_info_depth.results[:depth].to_i

        (depth - 1).downto(0).to_enum.map do |frame_n|
          proxy.stack_select_frame(frame_n)
          frame = proxy.stack_frame_info.results[:frame]
          locals = extract_locals
          args = extract_args(frame_n)

          {
            :lineno => frame.line,
            :filename => frame.file,
            :method => frame.func,
            :locals => locals,
            :args => args
          }
        end
      end

      def extract_locals
        proxy.stack_list_locals.results[:locals].reduce({}) do |acc, l|
          name = l[:name]
          value = l[:value]

          acc[name] = value

          acc
        end
      end

      def extract_args(frame_n)
        args_list = proxy.stack_list_arguments(frame_n, frame_n).results[:stack_args][0][:frame].args

        args_list.map { |arg| arg[:value] }
      end
    end
  end
end

File: builder.go
package main

import (
	"time"
	"fmt"
	"strconv"
)


func BuildPayload(gdb *Gdb) {
	gdb.Start()

	depth, _ := strconv.ParseInt(gdb.Send("-stack-info-depth")["payload"].(map[string]interface{})["depth"].(string), 10, 64)

	frames := make([]map[string]interface{}, depth)

	for frameNumber := depth - 1; frameNumber >= 0; frameNumber-- {
		gdb.Send(fmt.Sprintf("-stack-select-frame %d", frameNumber))
		frame := gdb.Send(fmt.Sprintf("-stack-info-frame"))["payload"].(map[string]interface{})["frame"]

		file := frame.(map[string]interface{})["file"]
		line := frame.(map[string]interface{})["line"]
		function := frame.(map[string]interface{})["func"]

		locals := extractLocals(gdb)
		args := extractArgs(gdb, frameNumber)

		frames[frameNumber] = map[string]interface{}{
			"filename": file,
			"lineno": line,
			"method": function,
			"locals": locals,
			"args": args,
		}
	}

	payload := map[string]interface{}{
		"access_token": ConfigurationAccessToken(),
		"data": map[string]interface{}{
			"timestamp": time.Now().Unix(),
			"environment": "development",
			"level": "error",
			"language": "c",
			"server": map[string]interface{}{
				"host": "my-host",
			},
			"notifier": map[string]interface{}{
				"name": "rollbar-dumps",
				"version": ConfigurationNotifierVersion(),
			},
			"body": map[string]interface{}{
				"trace": map[string]interface{}{
					"frames": frames,
					"exception": map[string]interface{}{
						"class": "MyError",
						"message": "The error message",
					},
				},
			},
		},
	}

	SendPayload(payload)
}


func extractLocals(gdb *Gdb) map[string]interface{} {
	gdbResult := gdb.Send("-stack-list-locals 1")
	locals := gdbResult["payload"].(map[string]interface{})["locals"].([]interface{})
	localsMap := make(map[string]interface{})

	for _, v := range(locals) {
		localVar := v.(map[string]interface{})

		localsMap[localVar["name"].(string)] = localVar["value"]
	}

	return localsMap
}


func extractArgs(gdb *Gdb, frameNumber int64) []interface{} {
	gdbResult := gdb.Send(fmt.Sprintf("-stack-list-arguments 1 %d %d", frameNumber, frameNumber))
	payload := gdbResult["payload"].(map[string]interface{})
	args := payload["stack-args"].([]interface{})[0].(
		map[string]interface{})["frame"].(
			map[string]interface{})["args"].(
				[]interface{})

	result := make([]interface{}, len(args))

	for index, arg := range(args) {
		result[index] = arg.(map[string]interface{})["value"]
	}

	return result
}
